### PR TITLE
[WIP] GotoDefinition Generates Metadata Signature File For External Symbols

### PR DESCRIPTION
**Pull request by cloudRoutine**
So far I've been able to implement the signature file generation function, save it to a unique disk location, and accurately locate the target symbol in the generated file. 

_sample `.fsi` generated for `XContainter`_
```fsharp
namespace System.Xml.Linq
open System.Xml.Linq

[<Class>]
type XContainer =
    inherit XNode
    /// Adds the specified content as children of this System.Xml.Linq.XContainer.
    /// content: A content object containing simple content or a collection of content objects to be added.
    member Add : content:obj -> unit
    /// Adds the specified content as children of this System.Xml.Linq.XContainer.
    /// content: A parameter list of content objects.
    member Add : content:obj [] -> unit
    /// Adds the specified content as the first children of this document or element.
    /// content: A content object containing simple content or a collection of content objects to be added.
    member AddFirst : content:obj -> unit
    /// Adds the specified content as the first children of this document or element.
    /// content: A parameter list of content objects.
    member AddFirst : content:obj [] -> unit
    /// Creates an System.Xml.XmlWriter that can be used to add nodes to the System.Xml.Linq.XContainer.
    member CreateWriter : unit -> System.Xml.XmlWriter
    /// Returns a collection of the descendant nodes for this document or element, in document order.
    member DescendantNodes : unit -> System.Collections.Generic.IEnumerable<XNode>
    /// Returns a collection of the descendant elements for this document or element, in document order.
    member Descendants : unit -> System.Collections.Generic.IEnumerable<XElement>
    /// Returns a filtered collection of the descendant elements for this document or element, in document order. Only elements that have a matching System.Xml.Linq.XName are included in the collection.
    /// name: The System.Xml.Linq.XName to match.
    member Descendants : name:XName -> System.Collections.Generic.IEnumerable<XElement>
    /// Gets the first (in document order) child element with the specified System.Xml.Linq.XName.
    /// name: The System.Xml.Linq.XName to match.
    member Element : name:XName -> XElement
    /// Returns a collection of the child elements of this element or document, in document order.
    member Elements : unit -> System.Collections.Generic.IEnumerable<XElement>
    /// Returns a filtered collection of the child elements of this element or document, in document order. Only elements that have a matching System.Xml.Linq.XName are included in the collection.
    /// name: The System.Xml.Linq.XName to match.
    member Elements : name:XName -> System.Collections.Generic.IEnumerable<XElement>
    /// Get the first child node of this node.
    member FirstNode : XNode
    /// Get the last child node of this node.
    member LastNode : XNode
    /// Returns a collection of the child nodes of this element or document, in document order.
    member Nodes : unit -> System.Collections.Generic.IEnumerable<XNode>
    /// Removes the child nodes from this document or element.
    member RemoveNodes : unit -> unit
    /// Replaces the children nodes of this document or element with the specified content.
    /// content: A content object containing simple content or a collection of content objects that replace the children nodes.
    member ReplaceNodes : content:obj -> unit
    /// Replaces the children nodes of this document or element with the specified content.
    /// content: A parameter list of content objects.
    member ReplaceNodes : content:obj [] -> unit
```

_symbol located @_ - `C:\Users\jared\Github\Owned\Forge\src\Forge.ProjectSystem\obj\3808760524\System.Xml.Linq.XContainer.fsi (5,5--5,15)`  

The part that I'm currently stuck on is getting this generated metadata file to register correctly as a part of the workspace so the document navigation service can locate it.

In my past experience  when working with Roslyn workspaces this would be a successful way to update one with a new document

```fsharp
            | FoundExternal (sourceDocument,ast,symbol) ->
                let! signatureText, sigPath,range = metadataService.TryFindMetadataRange(sourceDocument,ast,symbol)
                let workspace = document.Project.Solution.Workspace
                let sigSourceText = SourceText.From(signatureText,System.Text.Encoding.UTF8)
                let sigDocument = 
                    document.Project.AddDocument
                        (   Path.GetFileNameWithoutExtension sigPath
                        ,   signatureText
                        ,   folders=[Path.GetDirectoryName sigPath]
                        ,   filePath=sigPath 
                        )
                projectInfoManager.RegisterSignature sigPath sigDocument.Id
                let applyResult = workspace.TryApplyChanges (sigDocument.Project.Solution)
```

but the VisualStudioWorkspace seems to operate quite differently than the ones I'm used to and fails with

```
Exception thrown: 'System.InvalidOperationException' in Microsoft.VisualStudio.LanguageServices.dll
---- DEBUG ASSERTION FAILED ----
---- Assert Short Message ----
Unexpected exception seen in language service
---- Assert Long Message ----
System.InvalidOperationException: Operation is not valid due to the current state of the object.
   at Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioWorkspaceImpl.GetPreferredExtension(IVisualStudioHostProject hostProject, SourceCodeKind sourceCodeKind)
   at Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioWorkspaceImpl.AddDocumentToProjectItems(IVisualStudioHostProject hostProject, ProjectItems projectItems, DocumentId documentId, String folderPath, String documentName, SourceCodeKind sourceCodeKind, SourceText initialText, String filePath, Boolean isAdditionalDocument)
   at Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioWorkspaceImpl.AddDocumentToFolder(IVisualStudioHostProject hostProject, Project project, DocumentId documentId, IEnumerable`1 folders, String documentName, SourceCodeKind sourceCodeKind, SourceText initialText, String filePath, Boolean isAdditionalDocument)
   at Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioWorkspaceImpl.AddDocumentCore(DocumentInfo info, SourceText initialText, Boolean isAdditionalDocument)
   at Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioWorkspaceImpl.ApplyDocumentAdded(DocumentInfo info, SourceText text)
   at Microsoft.CodeAnalysis.Workspace.ApplyProjectChanges(ProjectChanges projectChanges)
   at Microsoft.CodeAnalysis.Workspace.TryApplyChanges(Solution newSolution, IProgressTracker progressTracker)
   at Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioWorkspaceImpl.TryApplyChanges(Solution newSolution, IProgressTracker progressTracker)
   at Microsoft.CodeAnalysis.Workspace.TryApplyChanges(Solution newSolution)
   at <StartupCode$FSharp-Editor>.$GoToDefinitionService.FindDefinitionsAsyncAux@101-7.Invoke(Tuple`3 _arg10)
   at Microsoft.VisualStudio.FSharp.Editor.Pervasive.Bind@143-1.Invoke(FSharpOption`1 _arg2)
   at Microsoft.FSharp.Control.AsyncBuilderImpl.cont@823.Invoke(a a)
   at Microsoft.VisualStudio.FSharp.LanguageService.Assert.Exception(Exception e)
   at Microsoft.VisualStudio.FSharp.Editor.CommonRoslynHelpers.computation@89-2.Invoke(Exception _arg1)
   at Microsoft.FSharp.Control.AsyncBuilderImpl.callA@839.Invoke(AsyncParams`1 args)
   at <StartupCode$FSharp-Core>.$Control.loop@426-38.Invoke(FSharpFunc`2 action)
   at Microsoft.FSharp.Control.Trampoline.ExecuteAction(FSharpFunc`2 firstAction)
   at Microsoft.FSharp.Control.TrampolineHolder.Protect(FSharpFunc`2 firstAction)
   at <StartupCode$FSharp-Core>.$Control.-ctor@509-1.Invoke(Object state)
   at System.Threading.QueueUserWorkItemCallback.WaitCallback_Context(Object state)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()
```

@JoshVarty suggested that I implement an additional internal workspace to track and manage the  metadatafiles like http://source.roslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/Implementation/MetadataAsSource/MetadataAsSourceFileService.cs,17

So that'll probably be the next thing I try to get around this issue.

@majocha  once that's done lets get this hooked up with https://github.com/Microsoft/visualfsharp/pull/2574 so we can click through to all symbols

### CODE REVIEW

File: vsintegration/src/FSharp.Editor/Common/Logging.fs [hunk 1]
```diff
@@ -69,6 +69,9 @@ type [<Export>] Logger [<ImportingConstructor>]
 
 [<AutoOpen>]
 module Logging =
+    open System.Diagnostics
```
Comment (saul) [inline vsintegration/src/FSharp.Editor/Common/Logging.fs]: Can we put these at the top of the file?

File: vsintegration/src/FSharp.Editor/Navigation/GoToDefinitionService.fs [hunk 2]
```diff
@@ -65,17 +82,37 @@ type internal FSharpGoToDefinitionService
...
+                let sigDocument = 
+                    document.Project.AddDocument
+                        (   Path.GetFileNameWithoutExtension sigPath
+                        ,   signatureText
+                        ,   folders=[Path.GetDirectoryName sigPath]
+                        ,   filePath=sigPath 
+                        )
+                projectInfoManager.RegisterSignature sigPath sigDocument.Id
+                let applyResult = workspace.TryApplyChanges (sigDocument.Project.Solution)
+                debug "Applied changes to workspace - %b" applyResult
```
Comment (saul) [inline vsintegration/src/FSharp.Editor/Navigation/GoToDefinitionService.fs]: Debug left in

File: vsintegration/src/FSharp.Editor/Navigation/GoToDefinitionService.fs [hunk 3]
```diff
@@ -65,17 +82,37 @@ type internal FSharpGoToDefinitionService
...
+                    let refTextSpan = CommonRoslynHelpers.FSharpRangeToTextSpan(refSourceText, range)
+                    results.Add(FSharpNavigableItem(refDocument, refTextSpan))
+                return results.AsEnumerable()
+            | FoundExternal (sourceDocument,ast,symbol) ->
+                let! signatureText, sigPath,range = metadataService.TryFindMetadataRange(sourceDocument,ast,symbol)
+                let workspace = document.Project.Solution.Workspace
+                let sigSourceText = SourceText.From(signatureText,System.Text.Encoding.UTF8)
+                let sigDocument = 
+                    document.Project.AddDocument
+                        (   Path.GetFileNameWithoutExtension sigPath
```
Comment (saul) [inline vsintegration/src/FSharp.Editor/Navigation/GoToDefinitionService.fs]: I don't think anywhere in the codebase uses this style for splitting long lines

File: vsintegration/src/FSharp.Editor/Navigation/NavigateToMetadataService.fs [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+                    match xmlIndex.GetMemberXML index with
+                    | VSConstants.S_OK, xml ->
+                        let processedXml = processXml xml
+                        let xmlDocs = ResizeArray ()
+                        match xmlIndexService.GetMemberDataFromXML processedXml with
+                        | VSConstants.S_OK, memberData ->
+                            match memberData.GetSummaryText () with
+                            | VSConstants.S_OK, xmlSummary -> xmlDocs.Add xmlSummary
+                            | _ -> ()
+                            match memberData.GetParamCount () with
```
Comment (saul) [inline vsintegration/src/FSharp.Editor/Navigation/NavigateToMetadataService.fs]: God this is deeply nested - is there no way we can flatten this out?

File: vsintegration/src/FSharp.Editor/Navigation/NavigateToMetadataService.fs [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+        | None -> []
+
+
+    /// Find the range of the target system in the generated signature metadata file
+    let tryFindExactLocation (filePath:string) (source:string) (currentSymbol:FSharpSymbolUse) (options:FSharpProjectOptions) sigCheck = asyncMaybe {
+            let checker = checkerProvider.Checker
+            let! symbolUses =
+                if not sigCheck then asyncMaybe{
+                    let! symbolUses = checker.GetAllUsesOfAllSymbolsInSourceString (options, filePath, source, false)  |> liftAsync
+                    debug "partial assmbly results for symbol - %A in '%s'"currentSymbol.Symbol.FullName filePath
```
Comment (saul) [inline vsintegration/src/FSharp.Editor/Navigation/NavigateToMetadataService.fs]: More debug lines - I'm not sure what the effect of these are

File: vsintegration/src/FSharp.Editor/Navigation/NavigateToMetadataService.fs [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+                    return! 
+                        None
+        }
+
+        
+    member __.TryFindMetadataRange (sourceDoc:Document, ast, fsSymbolUse) = async {
+        return! tryCreateMetadataContext sourceDoc ast fsSymbolUse
+    }
+
+                //context
```
Comment (saul) [inline vsintegration/src/FSharp.Editor/Navigation/NavigateToMetadataService.fs]: Commented code

**Comment by JoshVarty** [comment]
The exception you're seeing is because there are hard-coded checks for `.cs` and `.vb` in Roslyn:
http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs,808

I'm not really sure how to move forward. Might have to open an issue on Roslyn.

File: vsintegration/src/FSharp.Editor/Navigation/NavigateToMetadataService.fs (hunk 5 above)
Comment (cloudRoutine) [inline vsintegration/src/FSharp.Editor/Navigation/NavigateToMetadataService.fs]: it's like `printf` only it writes to the debug log. It only executes when the assembly is built in the debug configuration

**Comment by saul** [comment]
@JoshVarty - it's easy enough to open a pull request against dotnet/roslyn. I've done a couple and they're very accepting.

I imagine that this isn't the only place that we're going to hit this though. It would be easier to checkout Roslyn locally, make the changes, build it and deploy it to your FSharpDev VS instance. I did this a few weeks back and it worked surprisingly well.

**Comment by majocha** [comment]
@cloudRoutine Awesome! It will be easy to hookup if only we can map TypedAST `Entity` into ranges in the generated signature files.

**Comment by cloudRoutine** [comment]
@majocha you mean like this? 
https://github.com/cloudRoutine/visualfsharp/blob/47d7105d283ad793ac58832c73ee9fe8450e2967/vsintegration/src/FSharp.Editor/Navigation/NavigateToMetadataService.fs#L84-L174

**Comment by majocha** [comment]
@cloudRoutine I mean in my PR I simply bundle some data that is accessible at the moment of layout creation for the tooltip.[https://github.com/majocha/visualfsharp/blob/8af50af872a6cb2ec21d9880754f467593ba6d13/src/fsharp/TastOps.fs#L2510](https://github.com/majocha/visualfsharp/blob/8af50af872a6cb2ec21d9880754f467593ba6d13/src/fsharp/TastOps.fs#L2510)
I know next to nothing about all this, unfortunately. I'll try to read more in depth as soon as possible :)

**Comment by jasonmalinowski** [comment]
Just to chime in: VisualStudioWorkspace.TryApplyChanges means "modify my project to include these changes", which even if we ignore other Roslyn bug would mean you'd be adding the temporary file to the actual user's project...permanently. You will very much need to have a different workspace and not touch VisualStudioWorkspace here. If you look at the `MetadataAsSourceFileService` and squint a bit you'll see that although it has a bit that talks about symbols, only a bit of it does. A bit of refactoring might go a long way here in that code.

**Comment by cloudRoutine** [comment]
@jasonmalinowski good to know, it wasn't my intention to use that as the final implementation, I just wanted to get the workspace updated with the new doc added to the project to make sure the navigation service would successfully target the generated signature files.  

Do you have any recommendations about the best way to tie multiple workspaces together so that the navigation service reaches both of them and keeping them tracked by the same  the same opendocument table(service)? I'm really unsure about when I should be looking to the VS APIs for certain functionality, when I should be looking to the CodeAnalysis/Roslyn APIs, and when I just need to implement something myself; because it's hard to know what tools are already available. Ideally I'd like to use the CodeAnalysis/Roslyn APIs as much as possible as they're so much nicer than `IVsProjectFlavorUpgradeViaFactory2` et. al. 

Since the documentation for all of this is basically non-existent I've only been able to glean bits and pieces by reading the source, looking at other people's toy implementations and experimenting. My understanding and intuition about the topology of the system is lackluster at best. 😆  If you've got some useful heuristics about the best ways to get Roslyn and VS to play nice together they'd be much appreciated.

**Comment by smoothdeveloper** [comment]
> I imagine that this isn't the only place that we're going to hit this though.

Correct, I'm also looking for a way to open arbitrary script file from goto definition to complete the gotodef on `#load`, I hope I can reuse whatever we have to come up with for generated signature files.

File: vsintegration/src/FSharp.Editor/Navigation/NavigateToMetadataService.fs (hunk 4 above)
Comment (vasily-kirichenko) [inline vsintegration/src/FSharp.Editor/Navigation/NavigateToMetadataService.fs]: We need VSResult monad here ;)

**Comment by cloudRoutine** [comment]
@JoshVarty so I tried adding a new workspace just for sig metadata files and it seems like the new projects are being addeded to it and the workspace is update is working properly, but I still can't get the document to open properly.

**Comment by JoshVarty** [comment]
I'm fairly sure you have to do all of this painful stuff with the RDT and `SVsUIShellOpenDocument`: http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/Workspace/VisualStudioSymbolNavigationService.cs,114

My understanding of the `MetadataAsSourceFileService` is that it does roughly the following:

1. [Generate file on disk](http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/Workspace/VisualStudioSymbolNavigationService.cs,112) I think you're doing this properly.
2. [Open the file using `IVsUIShellOpenDocument` interface.](http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/Workspace/VisualStudioSymbolNavigationService.cs,118)
3. [Get the `ITextBuffer` for the now open document.](http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/Workspace/VisualStudioSymbolNavigationService.cs,122)
4. [Get the `Document` from the `ITextBuffer`](http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/Workspace/VisualStudioSymbolNavigationService.cs,134) I'm not sure if you have to do it this way, or if you could have held on to the original `Document` you created.
5. [`TryToNavigateToSpan()`](http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/Workspace/VisualStudioSymbolNavigationService.cs,140)

@jasonmalinowski Can probably correct me if I've misinterpreted these steps.

I've glanced through your code and I don't see anywhere in which the document will be opened by your source. As far as I know `TryNavigateToSpan` won't open documents for you, you have to do that with those `IVsUIShell` services.

**Comment by cloudRoutine** [comment]
it's finally opening the signature metadata!



what's left -
- code cleanup
- typecheck sig files properly
- open signatures as readonly
- improve speed  
 I think the VSShell calls might be running on the UI thread causing the slowdown as the task is running in the background

**Comment by cloudRoutine** [comment]
@saul then add some tests and get happy 😉

**Comment by saul** [comment]
I'll just leave this here...
https://github.com/Microsoft/visualfsharp/blob/master/CONTRIBUTING.md

😁

**Comment by cloudRoutine** [comment]
@saul right now I'm more concerned with usability so working on this repo stops being so miserable, i'll worry about stability after.

**Comment by saul** [comment]
Sure I'm not suggesting you drop everything and start writing tests, just that before it's merged it should have some 😀

**Comment by cloudRoutine** [comment]
Only one readonly window will open to show signature metadata, I tried a bunch of different ways to make it a preview/provisional texteditor, but nothing panned out.

**Comment by cloudRoutine** [comment]
well i sure made a mess of this 💀 

mulligan -> https://github.com/Microsoft/visualfsharp/pull/2658

**Pull request closed without merging**